### ISSUE TITLE
8080 version shows only E657

### ISSUE THREAD

**Issue by Ho-Ro** [WillStevens/basic1K#2]
Hi Will,
This is an ambitious project. I'm in the process of getting some milestones of 8080/Z80 Basic development to work in my [Z80 dongle](https://github.com/Ho-Ro/Z80_dongle) - Tiny Basic as one of the first open source projects, Nascom Basic as *my* first Basic at home - your project would fit as the smallest Basic. Unfortunately I cannot get it working correctly, your unchanged 8080 hex file responds to every input with `E657`, even direct after start:
```
>LIST

E657
```
I first suspected my setup, but exactly the same behaviour is seen with your [emulator](https://github.com/WillStevens/emulators).

The Z80 version does not respond, only the prompt `>` is visible.

**Comment by WillStevens** [comment]
Have you tried the release version? That's more likely to work then the 'HEAD' version in GitHub. The 8080 version won't work completely on a Z80, but it should do more than show E657.

**Comment by Ho-Ro** [comment]
#### CORRECTION
> I first suspected my setup, but exactly the same behaviour is seen with your [emulator](https://github.com/WillStevens/emulators).

This was my fault - sorry! The python emulator works as expected - IF the user is not so stupid to test it by typing only assignment statements like `A=1` as a test pattern :( `LET A=1` works, also `PRINT A` - shame on me.

---

I tried also the 8080 release version 1.0.0, this is the same as the inlined hex in the `emu8080.py`, it gives similar result E657 on my Z80 HW. I saw some references to parity in the source code that could yield different behaviour on a Z80.
Could you give me a hint for a working Z80 commit?

**Comment by Ho-Ro** [comment]
Finally I got the 8080 version working, there was an issue with `'\n'` and `'\r'` as line end on the linux side that I could fix w/o a SW change.

> The Z80 version does not respond, only the prompt > is visible.

Then I found a difference between the 8080 and Z80 versions, dunno if this is intended (at least it's confusing and not intuitive): 
The 8080 interprets status = 0 as no char available and jumps back on **Z** while the Z80 checks against 0x01 for no char available and jumps back on **NZ**.
The Z80 works for me if I change from `JR NZ` to `JR Z`.

### 8080
```asm
NextCharLoop:
	; This code is compatable with Stefan Tramm's
	; 8080 emulator
	IN 0
	ANA A
	RST_JZPage
	db (NextCharLoop&0ffh)-1
	IN 1
	MOV C,A
	OUT 1 ; echo
```

### Z80
```asm
NextCharLoop:
	; This code is compatable with Altair32
	; emulator
	IN A,(0)
	AND 1
	JR NZ,NextCharLoop
	IN A,(1)
	LD C,A
	OUT (1),A ; echo
```

**Comment by WillStevens** [comment]
Newlines and I/O status handling were often different on the different emulators that I was trying, so this section of code changed frequently. The tight code size limit and consequent fragility of the code to relocation compounded the issue. The Z80 version is incomplete - when I last worked on it I was trying to squeeze a few more bytes of space to fit in a few more commands. I’ll likely get back to this project again in the new year. In retrospect I wish that I had made the 8080 version Z80 compatible, but was able to save a byte or two at the time by using 8080 specific behaviour. I think I subsequently made it Z80 compatible, but there isn’t a release corresponding to this.

**Comment by Ho-Ro** [comment]
Hi Will,
I am closing this issue as it could be fixed from my side and `basicZ80` is now running in my [Z80 emulator](https://github.com/Ho-Ro/z80/tree/main/basic) as well as on my [HW](https://github.com/Ho-Ro/Z80_dongle).
During my 1st tests I found an issue with PEEK that can be fixed easily with #3.
Nevertheless, I would suggest documenting the HW dependencies, such as the addresses of the IO ports and the UART status bits, so that anyone can easily adapt the source code to the existing HW or emulator.
Currently: 
`basic8080`: IODATA=1, IOSTAT=0, IO char available: (IOSTAT) != 0
`basicZ80`: IODATA=1, IOSTAT=0, IO char available: (IOSTAT) & 01 == 0
Martin
